<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mathias">

<title>Análisis genómico comparativo de genes DGF-1 en Trypanosoma Cruzi</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="geneFeature_files/libs/clipboard/clipboard.min.js"></script>
<script src="geneFeature_files/libs/quarto-html/quarto.js"></script>
<script src="geneFeature_files/libs/quarto-html/popper.min.js"></script>
<script src="geneFeature_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="geneFeature_files/libs/quarto-html/anchor.min.js"></script>
<link href="geneFeature_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="geneFeature_files/libs/quarto-html/quarto-syntax-highlighting-fc6d358c97f25a8ea829b86655043430.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="geneFeature_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="geneFeature_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="geneFeature_files/libs/bootstrap/bootstrap-48972e06b888b3abdceb77e41c660eb4.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="geneFeature_files/libs/core-js-2.5.3/shim.min.js"></script>
<script src="geneFeature_files/libs/react-18.2.0/react.min.js"></script>
<script src="geneFeature_files/libs/react-18.2.0/react-dom.min.js"></script>
<script src="geneFeature_files/libs/reactwidget-2.0.0/react-tools.js"></script>
<link href="geneFeature_files/libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="geneFeature_files/libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="geneFeature_files/libs/reactable-0.4.4/reactable.css" rel="stylesheet">
<script src="geneFeature_files/libs/reactable-binding-0.4.4/reactable.js"></script>


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Análisis genómico comparativo de genes DGF-1 en <em>Trypanosoma Cruzi</em></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mathias </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="cell">
<style type="text/css">
p {
  text-align: justify
}

</style>
</div>
<section id="descripcion" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="descripcion"><span class="header-section-number">1</span> Descripcion</h2>
<p><img src="https://img.itch.zone/aW1nLzE3MTEyNjE5LmdpZg==/original/2%2BYVZg.gif" class="img-fluid" style="float:right;" alt="Illustration of a detective cat. Artwork by @https://shigemi.itch.io." width="301"></p>
<p><em>Trypanosoma cruzi</em>, el parásito causante de la enfermedad de Chagas, posee una familia de genes llamada <code>DGF-1</code> que juega un papel clave en su biología. En este estudio, realizamos un análisis comparativo de estos genes en distintos genomas para entender su diversidad y posibles funciones.</p>
<p>Los datos utilizados en este análisis fueron obtenidas a partir de diferentes genomas de <em>Trypanosoma cruzi</em>, disponibles en <a href="https://tritrypdb.org/tritrypdb/app"><strong>tritrydb</strong></a> y <a href="https://www.ncbi.nlm.nih.gov/"><strong>NCBI</strong></a>.</p>
<p>A partir de estos genomas, se extrajeron secuencias de los genes <code>DGF-1</code>, con el objetivo de realizar análisis comparativo. Inicialmente se estudió el contenido de <code>GC</code>, <code>longitud de las secuencias</code>, y otros aspectos relevantes para la caracterización de los genes de <code>DGF-1</code>.</p>
<section id="genomas" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="genomas"><span class="header-section-number">1.1</span> Genomas</h3>
<p>Este análisis utiliza los siguientes genomas <code>TriTrypDB-68_TcruziDm28c2018_Genome.fasta, TriTrypDB-68_TcruziBrazilA4_Genome.fasta, TriTrypDB-68_TcruziYC6_Genome.fasta, TcDm25_TcruziTcDm25H1_Genome.fasta</code>. A partir de los cuales se obtienen las secuencias para su estudio.</p>
</section>
<section id="librerias-y-datos-a-usar" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="librerias-y-datos-a-usar"><span class="header-section-number">1.2</span> Librerias y datos a usar</h3>
<p>Se cargan las <code>bibliotecas</code> necesarias para el análisis de datos genómicos, así como para la manipulación, transformación y visualización de dataframes, junto con la generación de gráficos en formato de arreglo. A continuación, se crea una lista de archivos que contiene los <code>genomas</code> a analizar, y se inicializa una <code>tabla</code> destinada a almacenar los resultados del análisis de los archivos FASTA.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cargar las librerías necesarias</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(seqinr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reactable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<hr>
</section>
</section>
<section id="procesamiento-de-los-genomas" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="procesamiento-de-los-genomas"><span class="header-section-number">2</span> Procesamiento de los genomas</h2>
<p>Para procesar las secuencias FASTA, se crea una función que genera un dataframe vacío donde se almacenarán los elementos obtenidos a partir de una secuencia dada. Luego, se elimina la ruta del archivo para extraer el nombre del mismo. A continuación, se obtiene el identificador de cada una de las secuencias dentro del archivo FASTA, se calcula la longitud de la secuencia y el contenido de GC utilizando funciones de la biblioteca SEQINR.</p>
<section id="características-de-los-genomas" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="características-de-los-genomas"><span class="header-section-number">2.1</span> Características de los genomas</h3>
<p>Se utiliza ALGO la función mencionada para extraer diversas características de los genomas. Con estos datos, generamos la primera <code>Tabla resumen</code>, que agrupa los contigs de cada genoma <code>TriTrypDB-68_TcruziDm28c2018_Genome.fasta, TriTrypDB-68_TcruziBrazilA4_Genome.fasta, TriTrypDB-68_TcruziYC6_Genome.fasta, TcDm25_TcruziTcDm25H1_Genome.fasta</code>, mostrando el <code>número total de secuencias encontradas</code>, así como la <code>longitud</code> mínima, media y máxima, así como el contenido de <code>GC</code> mínimo, medio y máximo.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Llamamos a la funcion para obtener GC y largo de cada contig, de cada genoma</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># genomes &lt;- calculate_fasta_feature(file.path("../../../study/maestria/genomic_data", transformed$fasta_basename), "genome")</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># genomes &lt;- transform_columns(genomes, "ID")</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># path_genome_safe_tsv &lt;- file.path("output_directory/", paste0("genome_gc_length.tsv"))</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># path_genome_safe &lt;- file.path("output_directory/", paste0("genome_gc_length.csv"))</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># write.csv(genomes, file = path_genome_safe, row.names = FALSE, col.names = TRUE, quote = FALSE)</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># write.table(genomes, file = path_genome_safe_tsv, row.names = FALSE, col.names = TRUE, sep = "\t", quote = FALSE)</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the table generated by the second script</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># genomes &lt;- read.table("output_directory/genome_gc_length.tsv", sep = "\t", header = TRUE)</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>genomes <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output_directory/genome_gc_length.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Crear una tabla con el conteo de secuencias por archivo</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>genomes_summary <span class="ot">&lt;-</span> genomes <span class="sc">%&gt;%</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(file) <span class="sc">%&gt;%</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Total_Sequences =</span> <span class="fu">n</span>(),</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_length =</span> <span class="fu">min</span>(Length),</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_length =</span> <span class="fu">round</span>(<span class="fu">mean</span>(Length), <span class="dv">0</span>),</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_length =</span> <span class="fu">max</span>(Length),</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_gc =</span> <span class="fu">min</span>(GC_Content),</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_gc =</span> <span class="fu">round</span>(<span class="fu">mean</span>(GC_Content), <span class="dv">3</span>),</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_gc =</span> <span class="fu">max</span>(GC_Content))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="tabla-resumen-de-los-genomas" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="tabla-resumen-de-los-genomas"><span class="header-section-number">2.2</span> Tabla resumen de los genomas</h3>
<div class="cell">
<div class="cell-output-display">
<div class="reactable html-widget html-fill-item" id="htmlwidget-d75f0caf784cb0da3bae" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-d75f0caf784cb0da3bae">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"file":["TcruziBrazilA4","TcruziDm28c2018","TcruziTcDm25H1","TcruziYC6"],"Total_Sequences":[402,636,35,266],"min_length":[703,911,47166,678],"mean_length":[113325,83761,1105091,177512],"max_length":[2738928,1645565,2824530,2951016],"min_gc":[0.415,0.363,0.242,0.445],"mean_gc":[0.533,0.519,0.505,0.524],"max_gc":[0.697,0.707,0.539,0.6]},"columns":[{"id":"file","name":"file","type":"character"},{"id":"Total_Sequences","name":"Total_Sequences","type":"numeric"},{"id":"min_length","name":"min_length","type":"numeric"},{"id":"mean_length","name":"mean_length","type":"numeric"},{"id":"max_length","name":"max_length","type":"numeric"},{"id":"min_gc","name":"min_gc","type":"numeric"},{"id":"mean_gc","name":"mean_gc","type":"numeric"},{"id":"max_gc","name":"max_gc","type":"numeric"}],"highlight":true,"dataKey":"02482dfea1094724c39d7bc52c115ef5"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="visualización-de-la-longitud-y-contenido-de-gc-de-los-genomas" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="visualización-de-la-longitud-y-contenido-de-gc-de-los-genomas"><span class="header-section-number">2.3</span> Visualización de la Longitud y Contenido de GC de los Genomas</h3>
<p>En esta sección, se generarán gráficos que representan el <code>contenido de GC</code> y la <code>longitud</code> de cada genoma. Se implementan funciones específicas para cada gráfico, lo que permite reutilizar el código de manera eficiente en otras partes del análisis. Para crear estas funciones se implementara las bibliotecas <code>ggridges</code>, <code>ggbeeswarm</code> y <code>cowplot</code>.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset nav-pills">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">GC</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">Length</a></li></ul>
<div class="tab-content nav-pills">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p>La función <code>plot_violin_gc</code> genera un gráfico de violín que muestra la distribución del contenido de GC en distintos genomas <code>TriTrypDB-68_TcruziDm28c2018_Genome.fasta, TriTrypDB-68_TcruziBrazilA4_Genome.fasta, TriTrypDB-68_TcruziYC6_Genome.fasta, TcDm25_TcruziTcDm25H1_Genome.fasta</code>.</p>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>genomes<span class="sc">$</span>Length <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>((<span class="fu">format</span>(genomes<span class="sc">$</span>Length, <span class="at">scientific =</span> <span class="cn">TRUE</span>))) <span class="co"># Changed format</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspired by http://tom-e-white.com/datavision/05-genome-size.html domain of https://github.com/tomwhite</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># GC plotting function</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>plot_violin_gc <span class="ot">&lt;-</span> <span class="cf">function</span>(data, plot_title, y_label) {</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> file, <span class="at">y =</span> GC_Content, <span class="at">fill =</span> file)) <span class="sc">+</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">width =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">fill =</span> <span class="st">"white"</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Dark2"</span>) <span class="sc">+</span>  <span class="co"># Apply RColorBrewer's Dark2 Palette</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(y_label) <span class="sc">+</span>  <span class="co"># Customize the ylab</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> plot_title,  <span class="co"># Customize the title</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">caption =</span> <span class="st">"Data source: Facultad de Ciencias, Udelar"</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal_vgrid</span>() <span class="sc">+</span> </span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.line.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>())</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Function plot_violin_gc call</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_violin_gc</span>(</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> genomes, </span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> <span class="st">"Genome GC"</span>, </span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">y_label =</span> <span class="st">"GC Content"</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-genome-gc-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y contenido GC.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>La función de R <code>plot_violin_length</code> crea un gráfico de violín que ilustra la distribución de las longitudes genómicas de los diferentes genomas <code>TriTrypDB-68_TcruziDm28c2018_Genome.fasta, TriTrypDB-68_TcruziBrazilA4_Genome.fasta, TriTrypDB-68_TcruziYC6_Genome.fasta, TcDm25_TcruziTcDm25H1_Genome.fasta</code>.</p>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to plot the length of genomes</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>plot_violin_length <span class="ot">&lt;-</span> <span class="cf">function</span>(data, type, plot_title, x_label) {</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">12345</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data, <span class="fu">aes</span>(Length, <span class="fu">reorder</span>(file, <span class="sc">-</span>Length, <span class="at">FUN =</span> median), <span class="at">color =</span> file)) <span class="sc">+</span> <span class="co"># reorder(file, -Length, FUN=median)</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_quasirandom</span>(<span class="at">groupOnX =</span> <span class="cn">FALSE</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>, <span class="at">size =</span> <span class="dv">1</span>, <span class="at">dodge.width =</span> <span class="fl">0.9</span>, <span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> plot_title,  <span class="co"># Title customization</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x_label,  <span class="co"># Custom x-axis label</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">caption =</span> <span class="st">"Data source: Facultad de Ciencias, Udelar"</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal_vgrid</span>() <span class="sc">+</span> </span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.line.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Dark2"</span>)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Conditional to choose the type of scale on the x axis</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (type <span class="sc">==</span> <span class="st">"genome"</span>) {</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_log10</span>(</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">10</span><span class="sc">^</span><span class="dv">2</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">3</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">4</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">5</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">6</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">7</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">8</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">9</span>, <span class="dv">10</span><span class="sc">^</span><span class="dv">10</span>),</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log10"</span>, scales<span class="sc">::</span><span class="fu">math_format</span>(<span class="dv">10</span><span class="sc">^</span>.x))</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Function plot_violin_length call</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_violin_length</span>(</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> genomes,</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">"genome"</span>, </span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> <span class="st">"Genome size"</span>, </span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">x_label =</span> <span class="st">"Number of base pairs"</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-genome-length-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y contenido GC.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="procesamiento-del-total-de-las-secuencias-codificantes" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="procesamiento-del-total-de-las-secuencias-codificantes"><span class="header-section-number">3</span> Procesamiento del total de las secuencias codificantes</h2>
<p>Se llama a la funcion <code>calculate_fasta_feature</code> para calcular en el total de <code>secuencia codificante</code> el contenido de <code>GC</code> y la <code>longitud (pb)</code>. Finalmente, creamos una <code>tabla resumen</code>, que agrupa el <code>total secuencias codificantes</code> por cada genoma <code>TriTrypDB-68_TcruziDm28c2018_Genome.fasta, TriTrypDB-68_TcruziBrazilA4_Genome.fasta, TriTrypDB-68_TcruziYC6_Genome.fasta, TcDm25_TcruziTcDm25H1_Genome.fasta</code>, mostrando así como la <code>longitud</code> mínima, media y máxima, así como el contenido de <code>GC</code> mínimo, medio y máximo.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Llamamos a la funcion para obtener GC y largo de cada contig, de cada genoma</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all_code_prot &lt;- calculate_fasta_feature(file.path(output_dir ,transformed$out_rest_gscissors), "gene")</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># all_code_prot &lt;- transform_columns(all_code_prot, "ID")</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co"># path_all_prot_safe_tsv &lt;- file.path("output_directory/", paste0("all_code_protein_gc_length.tsv"))</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># path_all_prot_safe &lt;- file.path("output_directory/", paste0("all_code_protein_gc_length.csv"))</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># write.csv(all_code_prot, file = path_all_prot_safe, row.names = FALSE, col.names = TRUE, quote = FALSE)</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># write.table(all_code_prot, file = path_all_prot_safe_tsv, row.names = FALSE, col.names = TRUE, sep = "\t", quote = FALSE)</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the table generated by the second script</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># all_code_prot_data &lt;- read.table("output_directory/all_code_protein_gc_length.tsv", sep = "\t", header = TRUE)</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>all_code_prot_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output_directory/all_code_protein_gc_length.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Crear una tabla con el conteo de secuencias por archivo</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>all_code_prot_summary <span class="ot">&lt;-</span> all_code_prot_data <span class="sc">%&gt;%</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(file) <span class="sc">%&gt;%</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Total_Sequences =</span> <span class="fu">n</span>(),</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_length =</span> <span class="fu">min</span>(Length),</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_length =</span> <span class="fu">round</span>(<span class="fu">mean</span>(Length), <span class="dv">0</span>),</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_length =</span> <span class="fu">max</span>(Length),</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_gc =</span> <span class="fu">min</span>(GC_Content),</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_gc =</span> <span class="fu">round</span>(<span class="fu">mean</span>(GC_Content), <span class="dv">3</span>),</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_gc =</span> <span class="fu">max</span>(GC_Content))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="tabla-resumen-de-secuencias-codificantes-por-genoma" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="tabla-resumen-de-secuencias-codificantes-por-genoma"><span class="header-section-number">3.1</span> Tabla resumen de secuencias codificantes por genoma</h3>
<div class="cell">
<div class="cell-output-display">
<div class="reactable html-widget html-fill-item" id="htmlwidget-5bf5f80fe868cc41ad09" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-5bf5f80fe868cc41ad09">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"file":["TcruziBrazilA4","TcruziDm28c2018","TcruziTcDm25H1","TcruziYC6"],"Total_Sequences":[14233,15319,18096,14268],"min_length":[102,156,7,141],"mean_length":[1264,1459,1356,1394],"max_length":[19908,15471,48501,19908],"min_gc":[0.277,0.277,0.278,0.249],"mean_gc":[0.541,0.531,0.542,0.54],"max_gc":[0.751,0.707,0.749,0.744]},"columns":[{"id":"file","name":"file","type":"character"},{"id":"Total_Sequences","name":"Total_Sequences","type":"numeric"},{"id":"min_length","name":"min_length","type":"numeric"},{"id":"mean_length","name":"mean_length","type":"numeric"},{"id":"max_length","name":"max_length","type":"numeric"},{"id":"min_gc","name":"min_gc","type":"numeric"},{"id":"mean_gc","name":"mean_gc","type":"numeric"},{"id":"max_gc","name":"max_gc","type":"numeric"}],"highlight":true,"dataKey":"12bef9b85fd5a27bdee2e6a5847dcb08"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset nav-pills">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true" href="">GC</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false" href="">Length</a></li></ul>
<div class="tab-content nav-pills">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<p>Llamamos a la funcion <code>plot_violin_gc</code>, dando lugar a un gráfico violín que muestra la distribución del contenido de GC en distintos genomas de las secuencias.</p>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Function plot_violin_gc call</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_violin_gc</span>(</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> all_code_prot_data, </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> <span class="st">"Coding sequence GC"</span>, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">y_label =</span> <span class="st">"GC Content"</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-gc%20all%20coding%20protein-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y contenido GC.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<p>Además, reutilizamos la función previamente empleada para graficar la distribución del genoma, pero esta vez generamos un gráfico que muestra la distribución de las longitudes de todas las secuencias codificantes, correspondientes a cada genoma.</p>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Function plot_violin_length call</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_violin_length</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> all_code_prot_data,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"genome"</span>, </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> <span class="st">"coding sequence size"</span>, </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x_label =</span> <span class="st">"Number of base pairs"</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-length%20all%20coding%20protein-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y largo total de la secuencia.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="procesamiento-de-las-secuencias-dgf-1" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="procesamiento-de-las-secuencias-dgf-1"><span class="header-section-number">4</span> Procesamiento de las secuencias <code>DGF-1</code></h2>
<p>La función <code>calculate_fasta_feature</code> se usa para calcular para cada secuencia el contenido de <code>GC</code> y la <code>longitud (pb)</code> de <code>DGF-1</code>. Se crea una <code>tabla resumen</code>, que agrupa las secuencias por cada genoma <code>TriTrypDB-68_TcruziDm28c2018_Genome.fasta, TriTrypDB-68_TcruziBrazilA4_Genome.fasta, TriTrypDB-68_TcruziYC6_Genome.fasta, TcDm25_TcruziTcDm25H1_Genome.fasta</code>, mostrando el <code>DGF-1</code>, así como la <code>longitud</code> mínima, media y máxima, así como el contenido de <code>GC</code> mínimo, medio y máximo.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>target_protein <span class="ot">&lt;-</span> <span class="fu">calculate_fasta_feature</span>(<span class="fu">file.path</span>(output_dir, transformed<span class="sc">$</span>out_gscissors_all), <span class="st">"gene"</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>target_protein <span class="ot">&lt;-</span> <span class="fu">transform_columns</span>(target_protein, <span class="st">"ID"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Crear una tabla con el conteo de secuencias por archivo</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>file_sequence_counts <span class="ot">&lt;-</span> target_protein <span class="sc">%&gt;%</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(file) <span class="sc">%&gt;%</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Total_Sequences =</span> <span class="fu">n</span>(),</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_length =</span> <span class="fu">min</span>(Length),</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_length =</span> <span class="fu">round</span>(<span class="fu">mean</span>(Length), <span class="dv">0</span>),</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_length =</span> <span class="fu">max</span>(Length),</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">min_gc =</span> <span class="fu">min</span>(GC_Content),</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_gc =</span> <span class="fu">round</span>(<span class="fu">mean</span>(GC_Content), <span class="dv">3</span>),</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">max_gc =</span> <span class="fu">max</span>(GC_Content))</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Mostrar la tabla de conteo de secuencias</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="fu">reactable</span>(file_sequence_counts, <span class="at">highlight =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="reactable html-widget html-fill-item" id="htmlwidget-66a54b29d82bb5a4b7f6" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-66a54b29d82bb5a4b7f6">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"file":["TcruziBrazilA4","TcruziDm28c2018","TcruziTcDm25H1","TcruziYC6"],"Total_Sequences":[44,75,192,102],"min_length":[267,7380,228,220],"mean_length":[6021,10334,5944,8811],"max_length":[10500,10821,10551,10521],"min_gc":[0.553,0.621,0.577,0.588],"mean_gc":[0.634,0.637,0.648,0.643],"max_gc":[0.698,0.644,0.691,0.688]},"columns":[{"id":"file","name":"file","type":"character"},{"id":"Total_Sequences","name":"Total_Sequences","type":"numeric"},{"id":"min_length","name":"min_length","type":"numeric"},{"id":"mean_length","name":"mean_length","type":"numeric"},{"id":"max_length","name":"max_length","type":"numeric"},{"id":"min_gc","name":"min_gc","type":"numeric"},{"id":"mean_gc","name":"mean_gc","type":"numeric"},{"id":"max_gc","name":"max_gc","type":"numeric"}],"highlight":true,"dataKey":"1558261d8e56c3170cd8af476364a4e4"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<section id="visualización-de-la-longitud-y-contenido-de-gc-de-dgf-1" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="visualización-de-la-longitud-y-contenido-de-gc-de-dgf-1"><span class="header-section-number">4.1</span> Visualización de la Longitud y Contenido de GC de <code>DGF-1</code></h3>
<p>Se generan gráficos que representan el <code>contenido de GC</code> y la <code>longitud</code> de cada secuencia de <code>DGF-1</code>. En esta seccion reutilizaremos parte del codigo escrito llamando a las funciones que crean los plot de <code>GC</code> y de <code>longitud</code> de las secuencias.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset nav-pills">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true" href="">GC</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false" href="">Length</a></li></ul>
<div class="tab-content nav-pills">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<p>Se usa la funcion <code>plot_violin_gc</code>, dando lugar a un gráfico violín que muestra la distribución del contenido de GC en distintos genomas de las secuencias <code>DGF-1</code>.</p>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>var_title_gc <span class="ot">&lt;-</span> <span class="fu">paste0</span>(key_name,<span class="st">" GC"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Function plot_violin_gc call</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_violin_gc</span>(</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> target_protein, </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> var_title_gc, </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">y_label =</span> <span class="st">"GC Content"</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-gc-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y contenido GC.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<p>Se usa la función previamente empleada para graficar la distribución del genoma, pero esta vez generamos un gráfico que muestra la distribución de las longitudes de las distintas secuencias de <code>DGF-1</code>, correspondientes a cada genoma.</p>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>var_title <span class="ot">&lt;-</span> <span class="fu">paste0</span>(key_name,<span class="st">" size"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Function plot_violin_length call</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_violin_length</span>(</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> target_protein,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"gene"</span>, </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> var_title, </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">x_label =</span> <span class="st">"Number of base pairs"</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>palmer_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-length-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y largo total de la secuencia.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="filtrado-de-las-secuencias-de-dgf-1-por-el-largo" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="filtrado-de-las-secuencias-de-dgf-1-por-el-largo"><span class="header-section-number">5</span> Filtrado de las secuencias de DGF-1 por el largo</h2>
<p>Se filtran las secuencias de DGF-1 de acuerdo a un largo mínimo (8000 pb). Las secuencias menores a este minimo se descartaran para este analisis inicial, y serán estudiadas para determinar si son pseudogenes.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># length_min &lt;- mean(target_protein$Length)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># length_min &lt;- median(file_sequence_counts$mean_length)</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># length_min &lt;- 8000</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># filter_data_higher &lt;- filter(target_protein, Length &gt;= length_min) </span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># filter_data_smaller &lt;- filter(target_protein, Length &lt;= length_min)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="resumen-de-información-de-dgf-1-completas-en-cada-genoma-filtradas-por-largo" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="resumen-de-información-de-dgf-1-completas-en-cada-genoma-filtradas-por-largo"><span class="header-section-number">5.1</span> Resumen de Información de DGF-1 completas en cada genoma (filtradas por largo)</h3>
<p>Se genera una tabla con las estadístias descriptivas de las DGF-1 completas</p>
<div class="cell">
<div class="cell-output-display">
<div class="reactable html-widget html-fill-item" id="htmlwidget-3d892e1a284a2d480314" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-3d892e1a284a2d480314">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"file":["TcruziBrazilA4","TcruziDm28c2018","TcruziTcDm25H1","TcruziYC6"],"Total_Sequences":[18,71,67,82],"min_length":[10056,10131,10003,9951],"mean_length":[10388,10412,10399,10401],"max_length":[10500,10821,10551,10521],"min_gc":[0.58,0.633,0.58,0.588],"mean_gc":[0.635,0.638,0.637,0.639],"max_gc":[0.644,0.644,0.643,0.645]},"columns":[{"id":"file","name":"file","type":"character"},{"id":"Total_Sequences","name":"Total_Sequences","type":"numeric"},{"id":"min_length","name":"min_length","type":"numeric"},{"id":"mean_length","name":"mean_length","type":"numeric"},{"id":"max_length","name":"max_length","type":"numeric"},{"id":"min_gc","name":"min_gc","type":"numeric"},{"id":"mean_gc","name":"mean_gc","type":"numeric"},{"id":"max_gc","name":"max_gc","type":"numeric"}],"highlight":true,"dataKey":"468467f842ec2baefea8abb3a718a92e"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="visualización-de-la-longitud-y-contenido-de-gc-de-dgf-1-completas" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="visualización-de-la-longitud-y-contenido-de-gc-de-dgf-1-completas"><span class="header-section-number">5.2</span> Visualización de la Longitud y Contenido de GC de <code>DGF-1</code> completas</h3>
<p>Se representan el <code>contenido de GC</code> y la <code>longitud</code> de cada secuencia de <code>DGF-1</code> completas (filtrada por un <code>largo minimo</code>).</p>
<div class="tabset-margin-container"></div><div class="panel-tabset nav-pills">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true" href="">GC</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false" href="">Length</a></li></ul>
<div class="tab-content nav-pills">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<p>Se usa la funcion <code>plot_violin_gc</code>, para representar la distribución del contenido de GC en distintos genomas de las secuencias <code>DGF-1</code> filtradas por un largo minimo.</p>
<div>

</div>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-GC-filter-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y largo total de la secuencia.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<p>Se usa la función <code>plot_violin_length</code> para representar distribución de las secuencias <code>DGF-1</code> filtradas.</p>
<div>

</div>
<div class="cell quarto-layout-panel" data-layout="[100]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="geneFeature_files/figure-html/plotting-length-filter-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Secuencias por genoma y largo total de la secuencia.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="resumen-de-información-de-secuencias-anotadas-como-dgf-1-menores-al-largo-minimo-establecido" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="resumen-de-información-de-secuencias-anotadas-como-dgf-1-menores-al-largo-minimo-establecido"><span class="header-section-number">6</span> Resumen de Información de secuencias anotadas como <code>DGF-1</code> menores al largo minimo establecido</h2>
<p>Se crea una nueva tabla resumen a partir de las secuencias que se descartaran por no cumplir con el minimo de largo establecido.</p>
<div class="cell">
<div class="cell-output-display">
<div class="reactable html-widget html-fill-item" id="htmlwidget-ac8ce9faa5d8f73cde0a" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac8ce9faa5d8f73cde0a">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"file":["TcruziBrazilA4","TcruziDm28c2018","TcruziTcDm25H1","TcruziYC6"],"Total_Sequences":[26,4,125,20],"min_length":[267,7380,228,220],"mean_length":[2997,8954,3556,2293],"max_length":[9410,9801,9835,9772],"min_gc":[0.553,0.621,0.577,0.597],"mean_gc":[0.634,0.63,0.654,0.658],"max_gc":[0.698,0.636,0.691,0.688]},"columns":[{"id":"file","name":"file","type":"character"},{"id":"Total_Sequences","name":"Total_Sequences","type":"numeric"},{"id":"min_length","name":"min_length","type":"numeric"},{"id":"mean_length","name":"mean_length","type":"numeric"},{"id":"max_length","name":"max_length","type":"numeric"},{"id":"min_gc","name":"min_gc","type":"numeric"},{"id":"mean_gc","name":"mean_gc","type":"numeric"},{"id":"max_gc","name":"max_gc","type":"numeric"}],"highlight":true,"dataKey":"09442c0906cf30339293d66da6d960f2"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<!-- ```{r}

filter_sequences_by_length <- function(fasta_files, min_length) {
  for (fasta_file in fasta_files) {
    # Leer el archivo FASTA
    sequences <- read.fasta(fasta_file, seqtype = "DNA", as.string = TRUE)
    
    # Filtrar secuencias con longitud mayor al mínimo
    filtered_seqs <- sequences[
      sapply(sequences, function(seq) {
        if (is.character(seq)) nchar(seq) else length(seq)
      }) >= min_length
    ]
    
    # Generar el nombre del archivo de salida
    output_file <- sub("\\.fasta$", paste0("_length", min_length, ".fasta"), fasta_file)
    
    # Guardar las secuencias filtradas si hay alguna
    if (length(filtered_seqs) > 0) {
      write.fasta(
        sequences = filtered_seqs,
        names = names(filtered_seqs),
        file.out = output_file
      )
      cat("Filtered sequences saved to:", output_file, "\n")
    } else {
      cat("No sequences longer than", min_length, "found in", fasta_file, "\n")
    }
  }
}
# Input: List of FASTA files and minimum sequence length
input_files <- fasta_files  # Replace with your file names
min_length <- 8000  # Set the desired minimum length

# Call the function
filter_sequences_by_length(input_files, min_length)

``` -->
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>